Per line history

samoylovfp opened the pull request:
This contains a lot of additional logic

samoylovfp committed:
Update Cargo.lock with the new version

samoylovfp committed:
Add a script for ramp-loading

samoylovfp committed:
Color each line with the history of cpu load

samoylovfp committed:
Clean up imports

samoylovfp commented on src/resample.rs:
```diff
@@ -0,0 +1,119 @@
+                resample(vec![0.0, 1.0, 2.0, 3.0], 14),
+                vec![
+                    0.0, 0.23076923, 0.46153846, 0.6923077, 0.9230769, 1.3076923, 1.7692308,
+                    2.2307692, 2.6923077, 2.2307692, 2.923077, 3.6153846, 4.3076925, 3.0
```
This is not correct

samoylovfp replied:
Fixed

samoylovfp committed:
Fix interpolation logic

samoylovfp commented on src/bin/ramp_load.rs:
```diff
@@ -0,0 +1,28 @@
+    );
+    loop {
+        for i in 0..cpu_num {
+            let load_duration = step_width.mul_f32((cpu_num - i) as f32 * 2.0);
```
This is actually an integer multiplication

samoylovfp replied:
Fixed

samoylovfp committed:
More edge case tests for resampling

samoylovfp commented:
![image](https://user-images.githubusercontent.com/17025459/209440789-1fc686ff-f171-4598-ac3f-0a2726ed1ee3.png)

demoth commented on src/bin/ramp_load.rs:
```diff
@@ -0,0 +1,32 @@
+/// Cpu load graph is supposed to look like a step-wise approximation of a triangle wave
+fn main() {
+    let step_width = Duration::from_millis(200);
+    let cpu_num = 24;
```
you can get it from sysinfo

samoylovfp replied:
Got it from there

demoth commented on src/resample.rs:
```diff
@@ -0,0 +1,128 @@
+        std::cmp::Ordering::Equal => source,
+
+        // Interpolate
+        std::cmp::Ordering::Less => {
```
can you please comment this block?

samoylovfp replied:
Done

demoth commented on src/resample.rs:
```diff
@@ -0,0 +1,128 @@
+        }
+
+        // Downsample
+        std::cmp::Ordering::Greater => {
```
can you please comment this block?

samoylovfp replied:
Done

demoth commented on src/resample.rs:
```diff
@@ -0,0 +1,115 @@
+
+    match source.len().cmp(&expected_len) {
+        // No work required
+        std::cmp::Ordering::Equal => source,
```
can it be `::More`?

samoylovfp commented:
I'll try to improve it tomorrow

samoylovfp committed:
Fix boolean comparison

demoth commented on src/main.rs:
```diff
@@ -56,20 +61,33 @@ fn print_colored(
+            stdout.set_color(ColorSpec::new().set_fg(Some(color)))?;
+            stdout.write_all(c.to_string().as_bytes())?;
+        }
+    } else {
```
else if cpu_load_samples.len() == 1 ?
else if cpu_loads.is_empty ?

samoylovfp replied:
cannot be empty, it is created with a single element on line 69.
If there is just a single element, then no processing is required

samoylovfp committed:
Move the "resample" module into the library crate

This is done for the possibility to run doc tests.
Rust has this limitation that only
doctests of public items in the library crate can be run.

After this change
lib.rs defines the "cpucat library crate",
main.rs defines the "cpucat binary crate"
and bin/ramp_load.rs defines the "ramp_load binary crate"

Also this allows generating documentation using `cargo doc`

samoylovfp committed:
Use interpolation for all resamplings

pub mod is required for the doctests

samoylovfp committed:
Get the number of cpu cores from sysinfo

demoth commented on src/lib.rs:
```diff
@@ -0,0 +1,3 @@
+pub mod resample_tools;
```
what this files does?

samoylovfp replied:
It tells the rust compiler to create a module called "resample_tools" from `src/resample_tools.rs` available from this library crate

samoylovfp replied:
Which is used for example in this doctest
https://github.com/demoth/cpucat/pull/3/files#diff-385473d0d79ed1ce67e90fea85b9f990680beefd4aa9bd3a2ba38b81e6bff1abR29-R37

samoylovfp replied:
Ah you mean the `lib.rs`? I tried to explain in the commit message. https://github.com/demoth/cpucat/pull/3/commits/8727c205e53265a3631cf1383eef568dba86f104
